### PR TITLE
Fix uClibc reallocarray compilation failure and also clang compilation

### PR DESCRIPTION
**Pull request by eaaltonen**
Build failure in http://autobuild.buildroot.net/results/9bda046c3af43765edbc27e2e3c9b6d4a635eef7 (issue #339) seems to be due to uClibc not implementing `reallocarray`. Use `realloc` instead and make the reason explicit.

Also included fix for making compile options clang friendly

### CODE REVIEW

**Comment by eaaltonen** [comment]
Alternative A) is the the commit "Use realloc with uClibc".

Alternative B) could be to use source comments about the reason, i.e.:
```
/* we use realloc, since uClibc does not implement reallocarray */
groups = realloc(groups, (grouplistsize * sizeof(gid_t)));
```

**Comment by eaaltonen** [comment]
There's also an option of adding a static check with `realloc` to make sure overflow does not occur, which would require compiling with C11 standard. This would be

`set_property(TARGET teeacl PROPERTY C_STANDARD 11)` to `libteeacl/CMakeLists.txt`
and
```
#include <assert.h>
#include <stdint.h>
#include <limits.h>
static_assert(NGROUPS_MAX < (SIZE_MAX / sizeof(gid_t)),
	"grouplist realloc could overflow");
```
to `libteeacl/src/group.c`.

**Comment by jforissier** [comment]
> Alternative B) could be to use source comments about the reason, i.e.:
> 
> ```
> /* we use realloc, since uClibc does not implement reallocarray */
> groups = realloc(groups, (grouplistsize * sizeof(gid_t)));
> ```

I like this option best. I don't think dealing with a possible overflow is relevant here.

**Comment by eaaltonen** [comment]
@etienne-lms : any preference?

**Comment by eaaltonen** [comment]
> > Alternative B) could be to use source comments about the reason, i.e.:
> > ```
> > /* we use realloc, since uClibc does not implement reallocarray */
> > groups = realloc(groups, (grouplistsize * sizeof(gid_t)));
> > ```
> 
> I like this option best. I don't think dealing with a possible overflow is relevant here.

Changed to B).

Alternative A) was macro based:
```
#ifdef __UCLIBC__
		groups = realloc(groups, (grouplistsize * sizeof(gid_t)));
#else
		groups = reallocarray(groups, grouplistsize, sizeof(gid_t));
#endif
```

**Review by etienne-lms — COMMENTED** [review]
Hello Eero,

Thanks for addressing this.


> > Alternative B) could be to use source comments about the reason, i.e.:
> > ```
> > /* we use realloc, since uClibc does not implement reallocarray */<
> > groups = realloc(groups, (grouplistsize * sizeof(gid_t)));
> > ```
> I like this option best. I don't think dealing with a possible overflow is relevant here.

That is also my preferred solution I didn't dare to propose. It simplifies libteeacl dependencies.
For commit "Use realloc instead of reallocarray", LGTM with a comment to address and minor changes in the commit message:
```patch
-Use realloc instead of reallocarray
+libteeacl: Use realloc() instead of reallocarray()
 
- Use `realloc` instead of `reallocarray`, since uClibc apparently does
- not implement `reallocarray`.
+ Use realloc() instead of reallocarray(), since uClibc apparently does
+ not implement reallocarray().
 
-Fixes #3399: `implicit declaration of function 'reallocarray'` build
-failure seen with GNU 11.3.0 C compiler and uClibc.
+Fixes: https://github.com/OP-TEE/optee_client/issues/339
```

For commit "Add -Wunsafe-loop-optimizations only for gcc": should this change be replicated in _flags.mk_ for Makefile driven builds?

**Comment by eaaltonen** [comment]
> For commit "Add -Wunsafe-loop-optimizations only for gcc": should this change be replicated in flags.mk for Makefile driven builds?

Yes. I did eventually find a way that works also in Makefile.

File: libteeacl/src/group.c [hunk 1]
```diff
@@ -72,7 +72,8 @@ enum rv_groupmember teeacl_user_is_member_of(const char *user, gid_t group)
 	if (ret == -1) {
 		p_groups = groups;
 
-		groups = reallocarray(groups, grouplistsize, sizeof(gid_t));
+		/* we use realloc, since uClibc does not implement reallocarray */
+		groups = realloc(groups, (grouplistsize * sizeof(gid_t)));
```
[the remark this answers is not in the document]
Comment (eaaltonen) [reply libteeacl/src/group.c]: > remove inner parentheses

done

> If allocation fails, old groups ref is lost and allocation done at line 66 is not freed. 

The groups ref is stored to `p_groups` when going into this branch, and freed if `realloc` fails, so I think we should be covered here.

**Comment by eaaltonen** [comment]
> minor changes in the commit message

applied.

**Comment by eaaltonen** [comment]
`Acked` and `Reviewed`, thanks.

**Pull request merged**